### ISSUE TITLE
ML-Benchmark prediction submission: [Prithvi-UNet]

### ISSUE THREAD

**Issue by midatm1234** [WCRP-CORDEX/ml-benchmark#14]
### Institution ID

JPL

### Institution Name

Jet Propulsion Laboratory

### Emulator identifier (`emulator_id`)

Prithvi-UNet

### Emulator Description
Our emulator, Prithvi-UNet, is a Prithvi-WxC-based statistical downscaling model for CORDEX-ML experiments. It maps coarse atmospheric predictors (u, v, q, t, z) at 850, 700, and 500 hPa, together **without static orography** to high-resolution precipitation (pr) and daily maximum temperature (tasmax). In the current configuration, the model uses a Prithvi-WxC encoder backbone (best_rmse_UNET_small.pt from IBM's granite-wxc repo) with a convolutional encoder-decoder downscaling head and pixel-shuffle upsampling.

Workflow
Compute training-set scalars for predictors and targets → preprocess/regrid data onto the training target grid → train/fine-tune the downscaling model → generate predictions. The saved scalar files are then reused during training and inference.

Preprocessing and normalization
Two preprocessing components are central to this emulator setup. First, the model uses saved normalization arrays for inputs_mean, inputs_std, targets_mean, and targets_std, and the model configuration points input_mu, input_sigma, target_mu, and target_sigma to those same files. The scalar-generation script computes these quantities channel-wise from the training data.

Second, the coarse predictors are regridded before training and inference so they are spatially aligned with the high-resolution target grid. Rather than describing this as part of the emulator architecture itself, it is more accurate to describe regridding as a preprocessing step in the CORDEX_ML workflow.

### Target scaling details

In the current v6 configuration, dynamic predictors use training-set channel normalization:

x_norm = (x - input_mu) / (input_sigma + eps)

For target variables, tasmax uses gridpoint z-score normalization:

tasmax_norm = (tasmax - tasmax_mu) / tasmax_sigma

Precipitation (pr) does not use z-score or log1p normalization in v6. It uses divide-only scaling by the training-set global 95th percentile, pr_95th, with target_mu(pr)=0 and target_sigma(pr)=pr_95th:

pr_norm = pr / pr_95th

The precipitation output is modeled with a hurdle formulation: an occurrence branch predicts wet probability and an amount branch predicts positive precipitation amount. The raw amount output, z_amount, is passed through softplus:

amount_norm = log(1 + exp(z_amount))

The wet probability is:

p_wet = sigmoid(z_wet)

The denormalized precipitation prediction is:

if p_wet >= 0.5:
    pr = pr_95th * amount_norm
else:
    pr = 0

During training, the precipitation-specific loss is the sum of an occurrence loss and an amount loss. The occurrence target is wet = 1 when pr > 0.01, otherwise wet = 0.

occurrence_loss = BCEWithLogits(z_wet, wet)

amount_target_norm = pr / pr_95th

amount_loss = SmoothL1(amount_norm, amount_target_norm), computed on wet pixels only

L_pr = occurrence_loss + amount_loss

The occurrence and amount loss weights are both 1.0.

### Hardware and Training Details

Fine-tuning with five epochs required about 30 hours, and inference
over the evaluation period of 20 years required approximately two hours on a system with four Nvidia A10
GPUs.

### Stochastic/Probabilistic Output

no

### Reference URL

https://arxiv.org/abs/2409.13598

### Repository for Reproducibility

https://github.com/midatm1234/granite-wxc/tree/CORDEX_ML

### Additional Notes

Our v2 submission addressed the following issues.

- NaN predictor values in the South Africa domain.
- Mishandling of the orography input in part of the workflow.

Our v3 submission includes the following updates.

- Improved smoothness in the predicted tasmax distributions
- Implementation of softplus normalization for pr

v4 Submission 

- Extended Fine-tuning: Increased training duration from 5 epochs (v3) to 10 epochs.

v5 Submission
- log1p precipitation normalization
- Batch size is 32 
- Loss function includes distribution differences

v6 Submission (current)
- Predictor normalization: x_norm = (x - input_mu) / (input_sigma + eps)
- tasmax uses gridpoint z-score normalization.
- pr uses divide-only scaling by the training-set global 95th percentile.
- pr is modeled with a hurdle head: wet occurrence probability plus positive amount prediction.
- Precipitation loss combines BCEWithLogits occurrence loss and SmoothL1 amount loss on wet pixels only.
- Wet threshold changed from 0.5 to 0.01; inference wet-probability threshold is 0.5.

**Comment by jgonzalezab** [comment]
@midatm1234 Unfortunately, due to an issue with the rsync synchronization on our end, some of the early submissions were corrupted. Would you mind uploading your submission again? You can continue to use the same format (`<emulator_id>.zip`) and same ssh key. We sincerely apologize for the inconvenience.

**Comment by midatm1234** [comment]
I’ve uploaded Prithvi-UNet.zip again as requested. Please let me know if there are any issues with the new submission.
Thank you for your help!

**Comment by jgonzalezab** [comment]
Hi @midatm1234 . We’re currently unable to locate the` Prithvi-UNet.zip` file on our cluster. Could you please confirm that it was transferred successfully? If not, would you mind resending it again?

**Comment by midatm1234** [comment]
I sent the file twice using this command:
rsync -avP --rsh="ssh -i /Users/huikyole/.ssh/id_rsa" Prithvi-UNet.zip anonsftp@ui.sci.unican.es::abadj-rw
Could you please double-check whether there’s anything wrong with it?

File transfer has been completed.

**Comment by jgonzalezab** [comment]
We got the submission, thanks! I’ve been going through them and everything looks good so far. Just one small thing, it looks like you uploaded two submissions both with and without orography. Could you open a new issue for the orography submission (`Prithvi-UNet-orog`) so we can register it properly? Thanks!!

**Comment by midatm1234** [comment]
Thanks for the update and for reviewing the submission!

Sure. I will open a new issue and submit two separate zip files so that the orography (Prithvi-UNet-orog) and no-orography submissions can be registered properly.

While double-checking the results, I also identified a couple of issues in my previous submission:

- Some NaN values in the South Africa domain.
- Mishandling of the orography input in part of the workflow.

I have corrected both issues. 
Would it be okay if I resubmit both the orog and no-orog predictions to ensure consistency and correctness across the benchmark?

Thanks again for your careful review.

**Comment by jgonzalezab** [comment]
Sure! You can submit both with (`Prithvi-UNet-orog-v2.zip`) and without (`Prithvi-UNet-v2.zip`). Please note the `v2` in the zip filename, Thanks!!

**Comment by midatm1234** [comment]
@jgonzalezab , both Prithvi-UNet-v3.zip and Prithvi-UNet-orog-v3.zip have been submitted. Could you please check if everything was received correctly? Thanks a lot.

**Comment by jgonzalezab** [comment]
Hi @midatm1234 . We noticed that you have submitted several additional versions of the model (both with and without orog), the latest being the v6 version transferred to the cluster around April 15. Could you provide us with more details about these new submissions? Should we consider them for the final version of the benchmark? Thanks!

**Comment by midatm1234** [comment]
@jgonzalezab Sure. Please use v6_orog and v6 for the final benchmark. These are the latest versions transferred to the cluster around April 15 and should replace the earlier submissions.

The update is described above as I have updated the main issue whenever submitting a new version. 
Please let me know if you need any additional metadata or clarification about the files.
